engines/sherlock/talk.cpp

```diff
@@ -0,0 +1,1740 @@
+					endStr) {
+				wait = 1;
+			}
+
+			switch (str >= scriptEnd ? 0 : str[0]) {
+			case SWITCH_SPEAKER:
+			case ASSIGN_PORTRAIT_LOCATION:
+			case BANISH_WINDOW:
+			case IF_STATEMENT:
+			case ELSE_STATEMENT:
+			case END_IF_STATEMENT:
+			case GOTO_SCENE:
+			case CALL_TALK_FILE:
+				wait = 1;
+				break;
+			default:
+				break;
+			}
+		}
+
+		// Open window if it wasn't already open, and text has already been printed
+		if ((openTalkWindow && wait) || (openTalkWindow && str[0] >= 128 && str[0] != CARRIAGE_RETURN)) {
+			if (!ui._windowStyle) {
+				screen.slamRect(Common::Rect(0, CONTROLS_Y, SHERLOCK_SCREEN_WIDTH, SHERLOCK_SCREEN_HEIGHT));
+			} else {
+				ui.summonWindow();
+			}
+
+			ui._windowOpen = true;
+			openTalkWindow = false;
+		}
+
+		if (wait) {
+			// Handling pausing
+			if (!pauseFlag && charCount < 160)
+				charCount = 160;
+
+			wait = waitForMore(charCount);
+			if (wait == -1)
+				endStr = true;
+
+			// If a key was pressed to finish the window, see if further voice files should be skipped
+			if (wait >= 0 && wait < 254) {
+				if (str[0] == SFX_COMMAND)
+					str += 9;
+			}
+
+			// Clear the window unless the wait was due to a PAUSE command
+			if (!pauseFlag && wait != -1 && str < scriptEnd && str[0] != SFX_COMMAND) {
+				if (!_talkStealth)
+					ui.clearWindow();
+				yp = CONTROLS_Y + 12;
+				charCount = line = 0;
+			}
+
+			pauseFlag = false;
+		}
+	} while (!_vm->shouldQuit() && !endStr);
+
+	if (wait != -1) {
+		for (int ssIndex = 0; ssIndex < (int)_savedSequences.size(); ++ssIndex) {
+			SequenceEntry &seq = _savedSequences[ssIndex];
+			Object &object = scene._bgShapes[seq._objNum];
+
+			for (uint idx = 0; idx < seq._sequences.size(); ++idx)
+				object._sequences[idx] = seq._sequences[idx];
+			object._frameNumber = seq._frameNumber;
+			object._seqTo = seq._seqTo;
+		}
+
+		pullSequence();
+		if (_speaker >= 0 && _speaker < 128)
+			people.clearTalking();
+	}
+}
+
+/**
+ * When the talk window has been displayed, waits a period of time proportional to
+ * the amount of text that's been displayed
+ */
+int Talk::waitForMore(int delay) {
+	Events &events = *_vm->_events;
+	People &people = *_vm->_people;
+	Scene &scene = *_vm->_scene;
+	Sound &sound = *_vm->_sound;
+	UserInterface &ui = *_vm->_ui;
+	CursorId oldCursor = events.getCursor();
+	int key2 = 254;
+
+	// Unless we're in stealth mode, show the appropriate cursor
+	if (!_talkStealth) {
+		events.setCursor(ui._lookScriptFlag ? MAGNIFY : ARROW);
+	}
+
+	do {
+		if (sound._speechOn && !*sound._soundIsOn)
+			people._portrait._frameNumber = -1;
+
+		scene.doBgAnim();
+
+		// If talkTo call was done via doBgAnim, abort out of talk quietly
+		if (_talkToAbort) {
+			key2 = -1;
+			events._released = true;
+		} else {
+			// See if there's been a button press
+			events.setButtonState();
+
+			if (events.kbHit()) {
+				Common::KeyState keyState = events.getKey();
+				if (keyState.keycode >= 32 && keyState.keycode < 128)
```

Comment:
This looks a bit odd to compare key codes against numbers. Shouldn't this rather be constant names? Then again, maybe there is a specific reason why these are just numbers and using `keycode` might not be the way to go?